src/import-shadow-hash.c

```diff
@@ -0,0 +1,173 @@
+  // Create a new Cache Entry.
+  CacheEntry *entry = CacheEntryNew();
+  GBytes* gbytes = g_bytes_new(sdhash, strlen(sdhash));
+  // Set the newly created entry hash using the /etc/shadow hash.
+  CacheEntryHashSet(entry, CACHE_ENTRY_ALGORITHM_CRYPT, gbytes, NULL);
+  g_bytes_unref(gbytes);
+  GError *error = NULL;
+  CacheStorage *storage = CacheStorageNew(DEFAULT_STORAGE_PATH);
+  if (!CacheStoragePutEntry(storage, username, entry, &error))
+      goto done;
+
+  done:
+    if (error) {
+      fprintf(stderr, "Failed to create Cache Entry.\n");
+      return FALSE;
+    } else {
+      fprintf(stdout, "Successfully created Cache Entry.\n");
+      return TRUE;
+    }
+}
+
+/**
+ * UpdateShadowHash:
+ * @username: Username to blank out hashed shadow password.
+ * @towhat: What to changed the hashed password to.
+ */
+gboolean UpdateShadowHash(const gchar *username, gchar *towhat) {
+  struct spwd *stmpent = NULL;
+
+  FILE *pwfile, *opwfile;
+  int error = 0;
+  int oldmask = umask(077);
+  int wroteentry = 0;
+
+
+  pwfile = fopen(SH_TMPFILE, "w");
+  umask(oldmask);
+  if (pwfile == NULL) {
+    error = 1;
+    goto done;
+  }
+
+  opwfile = fopen("/etc/shadow", "r");
+  if (opwfile == NULL) {
+    fclose(pwfile);
+    error = 1;
+    goto done;
+  }
+
+  // stmpent -- Read next shadow entry from STREAM.
+  stmpent = fgetspent(opwfile);
+  // Loop through /etc/shadow file
+  while (stmpent) {
+    if (!strcmp(stmpent->sp_namp, username)) {
+      // Create entry from shadow hash if username matches an entry.
+      CreateEntryFromShadowHash(username, stmpent->sp_pwdp);
+      // Set encrypted password to towhat
+      stmpent->sp_pwdp = towhat;
+      stmpent->sp_lstchg = time(NULL) / (60 * 60 * 24);
+      // Check that an entry was written.
+      wroteentry = 1;
+      fprintf(stdout, "Set password %s for %s\n", stmpent-> sp_pwdp, username);
+    }
+
+    if (putspent(stmpent, pwfile)) {
+      fprintf(stderr, "Error writting entry to shadow file\n");
+      error = 1;
+      break;
+    }
+
+    stmpent = fgetspent(opwfile);
+  }
+
+  fclose(opwfile);
+
+  done:
+    if (!error) {
+      // Move tmp file contents into the real file.
+      if (!rename(SH_TMPFILE, "/etc/shadow"))
+          fprintf(stdout, "Password changed for %s\n", username);
+      else
+          error = 1;
+    }
+
+  if (!error) {
+    fprintf(stdout, "Successfully updated shadow file.\n");
+    return TRUE;
+  } else {
+    unlink(SH_TMPFILE);
+    fprintf(stderr, "Failed to update shadow file.\n");
+    return FALSE;
+  }
+}
+
+int main(int argc, char *argv[]) {
+  int exit_code = 0;
+  GError *error = NULL;
+  GOptionContext *context = NULL;
+ 
+  // TODO(mcclunge): Add option to disable hash password instead of just remove.
+  static gboolean remove = FALSE; 
+  static gchar* username = NULL;
+  static GOptionEntry entries[] = 
+    {
+	{ "remove", 'r', 0, G_OPTION_ARG_NONE, &remove, "Remove shadow hash password for a given user."},
```

Comment:
Instead of removing the old shadow entry immediately, we want to lock it for the first phase of the rollout and unlock it if a rollback is needed. It might be wise to make this file simply read /etc/shadow and add a cache entry instead of actually modifying /etc/shadow. Then we can do the /etc/shadow modification with "passwd --lock", "passwd --unlock", etc.